GVFS/GVFS.Common/NuGetUpgrade/NuGetUpgrader.cs

```diff
@@ -492,6 +522,66 @@ private bool TryRecursivelyDeleteInstallerDirectory(out string error)
             return true;
         }
 
+        private IList<IPackageSearchMetadata> QueryFeedWithRetry()
+        {
+            int retryCount = 0;
+            IList<IPackageSearchMetadata> queryResults = null;
+
+            do
+            {
+                try
+                {
+                    queryResults = this.nuGetFeed.QueryFeedAsync(this.nuGetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+                    return queryResults;
+                }
+                catch (Exception ex) when (ex is FatalProtocolException &&
+                                           ex.InnerException != null &&
+                                           ex.InnerException is System.Net.Http.HttpRequestException &&
+                                           ex.InnerException.Message.Contains("401 (Unauthorized)") &&
+                                           retryCount == 0)
+                {
+                    EventMetadata data = CreateEventMetadata(ex);
+                    this.tracer.RelatedWarning(data, "Failed to query feed due to unauthorized error. Re-acquiring new credentials and trying again.");
+                    retryCount++;
+                    if (!this.TryRefreshCredentials(out string error))

```

Comment:
I think it would be relatively infrequent. It would indicate an error with permissions at that point, I think, such as lack of read access to the feed.

Reply:
I expect the `TryRefreshCredentials` to generally succeed, even if there is an initial failure in `QueryFeedAsync`.